### PR TITLE
ReplaceRefactoring: improve performance

### PR DESCRIPTION
**Pull request by jukzi**
Get rid of O(n^2) compares during search&replace with hashing.

### CODE REVIEW

File: org.eclipse.search/search/org/eclipse/search/internal/ui/text/ReplaceRefactoring.java [hunk 1]
```diff
@@ -290,31 +290,36 @@ public boolean hasMatches() {
...
-				matches.add(match);
-				return false;
+		Object key; // IFileStore or URI, see URIUtil.equals(URI, URI)
+		try {
+			key = EFS.getStore(uri);
+		} catch (CoreException e) {
+			// fall back to default equality test
+			key = uri;
+		}
+		if (fAlreadyCollected.containsKey(key)) {
```
Comment (iloveeclipse) [inline org.eclipse.search/search/org/eclipse/search/internal/ui/text/ReplaceRefactoring.java:309]: But that doesn't match what URIUtil does - it converts arguments before doing equals to `EFS.getStore(one)`. Here only the key is converted to IFileStore or URI, but so we compare apples with oranges?

Comment (jukzi) [reply org.eclipse.search/search/org/eclipse/search/internal/ui/text/ReplaceRefactoring.java:309]: see line 304: i convert too. the map then stores both apples and oranges in the bin (line 322+323). So i can test if either apple or orange is inside.

Comment (iloveeclipse) [reply org.eclipse.search/search/org/eclipse/search/internal/ui/text/ReplaceRefactoring.java:309]: Arrgh, I've overlooked that. 
Please add comment to the map definition that keys can be both apples and oranges.

Comment (iloveeclipse) [reply org.eclipse.search/search/org/eclipse/search/internal/ui/text/ReplaceRefactoring.java:309]: Beside this, LGTM

Comment (jukzi) [reply org.eclipse.search/search/org/eclipse/search/internal/ui/text/ReplaceRefactoring.java:312]: done

**Review by iloveeclipse — APPROVED** [review]
Thanks

**Pull request merged**